Feature-task3:

AlexLukashevich95 opened the pull request:
-changed controller to restController
-added ExceptionHandler for RestController

AlexLukashevich95 committed:
Feature-task3:
-changed controller to restController
-added ExceptionHandler for RestController

Evgnanii commented on src/main/java/com/alex/tasktable/exceptions/ControllerExceptionHandler.java:
```diff
@@ -0,0 +1,29 @@
+public class ControllerExceptionHandler {
+    @ExceptionHandler(TaskException.class)
+    @ResponseStatus(HttpStatus.NOT_FOUND)
+    public ResponseEntity<String> handleResourceNotFoundException(TaskException ex) {
```
Создай Dto  с полями. код ошибки, текст ошибки, название исключения

AlexLukashevich95 committed:
Feature-task3:
-divided TaskController into two classes
-added new Exceptions

Evgnanii commented on src/main/java/com/alex/tasktable/controller/TaskDataController.java:
```diff
@@ -0,0 +1,33 @@
+
+@RestController
+@RequestMapping("/tasks")
+public class TaskDataController {
```
Just TaskController

Evgnanii commented on src/main/java/com/alex/tasktable/controller/TaskDataController.java:
```diff
@@ -0,0 +1,33 @@
+    @Autowired
+    private TaskService taskService;
+
+    @PostMapping("/save")
```
no need /save here. https://restfulapi.net/resource-naming/

Evgnanii commented on src/main/java/com/alex/tasktable/controller/TaskDataController.java:
```diff
@@ -0,0 +1,33 @@
+    private TaskService taskService;
+
+    @PostMapping("/save")
+    public ResponseEntity<Task> createTask(@RequestBody Task task) throws TaskException {
```
cant throws TaskException. exception must be handeled

Evgnanii commented on src/main/java/com/alex/tasktable/controller/TaskPageController.java:
```diff
@@ -8,59 +8,50 @@
-    @RequestMapping(value = "/taskform")
-    public String showform(Model model) {
+    //@RequestMapping(value = "/taskform")
+    @GetMapping("/taskform")
```
it must be part of TaskController. in this class only main page return

Evgnanii commented on src/main/java/com/alex/tasktable/controller/TaskPageController.java:
```diff
@@ -8,59 +8,50 @@
 
-    @RequestMapping(value = "/save", method = RequestMethod.POST)
-    public String createTask(@ModelAttribute("task") Task task, Model model) {
+    //@RequestMapping(value = "/viewtask", method = RequestMethod.GET)
```
again, must be part of Rest Controller

Evgnanii commented on src/main/java/com/alex/tasktable/controller/TaskPageController.java:
```diff
@@ -8,59 +8,50 @@
+    public String getTasks(Model model) {
         try {
-            taskService.save(task);
+            model.addAttribute("list", taskService.findAll().stream().map(task -> taskMapper.toDto(task))
```
taskService must return dto

Evgnanii commented on src/main/java/com/alex/tasktable/mapper/TaskMapper.java:
```diff
@@ -9,8 +9,8 @@
 public class TaskMapper {
 
     private final DateTimeFormatter dateTimeFormatter = DateTimeFormatter.ofPattern("dd.MM.yyyy HH:mm");
```
try to use mapstruct

AlexLukashevich95 committed:
Feature-task3:
-delete unnecessary controller
-added new exception
-added new dto class

Evgnanii commented on src/main/java/com/alex/tasktable/controller/TaskController.java:
```diff
@@ -0,0 +1,83 @@
+    private TaskService taskService;
+
+    @Autowired
+    private TaskMapper taskMapper;
```
Сервис должен возвращать dto. Контроллер не должен заниматься маппингом,соответсвенно маппера тут быть не должно

Evgnanii commented on src/main/java/com/alex/tasktable/controller/TaskController.java:
```diff
@@ -0,0 +1,83 @@
+    private TaskMapper taskMapper;
+
+    @GetMapping("")
+    public String showMain(Model model) {
```
это не тут

Evgnanii commented on src/main/java/com/alex/tasktable/controller/TaskController.java:
```diff
@@ -0,0 +1,83 @@
+
+    @PutMapping("/{id}")
+    public ResponseEntity<TaskDto> updateTask(@PathVariable Long id, @RequestBody Task task) {
+        Optional.of(taskService.update(task)).orElseThrow(() -> new ResourceNotFoundException("Not found Task with id = " + id));
```
валидация и выборос ошибьки в сервисе должен быть

Evgnanii commented on src/main/java/com/alex/tasktable/controller/TaskController.java:
```diff
@@ -0,0 +1,83 @@
+        return "taskform";
+    }
+
+    @GetMapping("/viewtask")
```
у тебя же есть правильный @PutMapping для апдейта, зачем тут это?

Evgnanii commented on src/main/java/com/alex/tasktable/controller/TaskController.java:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    @GetMapping("/taskform")
+    public String showForm(Model model) {
```
Это не рест. Этот метод должен быть на taskPageController и  в нем не должно быть данных про задаче. В странице должен быть код на javascript ajax, который сделает get запрос на сервер и получит таски, чтобы их отобразить

Evgnanii commented on src/main/java/com/alex/tasktable/controller/TaskPageController.java:
```diff
@@ -1,68 +0,0 @@
-
-@Controller
-@RequestMapping("/tasks")
-public class TaskPageController {
```
верни этот контроллер для отображения страницы

Evgnanii commented on src/main/java/com/alex/tasktable/repository/impl/TaskRepositoryImpl.java:
```diff
@@ -60,12 +62,12 @@ public List<Task> findAll() throws TaskException {
             return tasks;
         } catch (SQLException e) {
-            throw new TaskException(e);
+            throw new BadRequestException("f");
```
Press f for respect?)

AlexLukashevich95 committed:
Feature-task3:
-replace methods which work with front to TaskPageController
-change location for calling resourcenotfoundexception

AlexLukashevich95 committed:
Feature-task3:
-add new jsp form
-changed sql exception handling(in process)

AlexLukashevich95 committed:
Feature-task3:
-fixed issues

AlexLukashevich95 committed:
Feature-task3:
-fixed issues

Evgnanii commented on src/main/java/com/alex/tasktable/controller/TaskController.java:
```diff
@@ -43,6 +43,7 @@ public ResponseEntity<Task> createTask(@RequestBody Task task) {
     @PutMapping("/{id}")
     public ResponseEntity<TaskDto> updateTask(@PathVariable Long id, @RequestBody Task task) {
         try {
+            task.setId(id);//чувствую, что надо сетать в jsp, но решил сделать так
```
Метод update сервиса принимает на вход id и dto

Evgnanii replied:
Поправь это и работу с exception и можно мержить, бери некст таску

AlexLukashevich95 committed:
Feature-task3:
-fixed issues
